### ISSUE TITLE
Integer bounds in quantifiers and pure functions

### ISSUE THREAD

**Issue by vakaras** [viperproject/prusti-dev#263]
Even when overflow checking is enabled, integer bounds are not known in quantifiers and pure functions. For this example:
```rust
use prusti_contracts::*;

#[pure]
#[ensures(result)]
fn test2(x: u32) -> bool { x >= 0 }

#[pure]
#[requires(_x >= 0)]
fn check(_x: u32) -> bool { true }

#[ensures(forall(|k: u32| check(k)))]
pub fn test() {}

fn main() {}
```
Prusti reports:
```plain
|__)  _\/_  |__) |  | /__`  |   ____\/_  | 
 |      /\   |  \ \__/ .__/  |       /\   | 

{commit e4907135 2020-11-20 12:07:41 UTC, built on 2020-11-21 16:13:41 UTC}


warning: function is never used: `test2`
 --> /home/data/eth/rust/prusti-dev/prusti-tests/tests/verify_overflow/pass/competitve_programming/dp/TheJackpot_10684_failing.rs:7:4
  |
7 | fn test2(x: u32) -> bool { x >= 0 }
  |    ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: comparison is useless due to type limits
 --> /home/data/eth/rust/prusti-dev/prusti-tests/tests/verify_overflow/pass/competitve_programming/dp/TheJackpot_10684_failing.rs:7:28
  |
7 | fn test2(x: u32) -> bool { x >= 0 }
  |                            ^^^^^^
  |
  = note: `#[warn(unused_comparisons)]` on by default

warning: comparison is useless due to type limits
  --> /home/data/eth/rust/prusti-dev/prusti-tests/tests/verify_overflow/pass/competitve_programming/dp/TheJackpot_10684_failing.rs:10:12
   |
10 | #[requires(_x >= 0)]
   |            ^^^^^^^

Verification of 4 items...
error: [Prusti internal error] unregistered verification error: [postcondition.violated:assertion.false] Postcondition of m_TheJackpot_10684_failing$$test2__$TY$__$int$$$bool$ might not hold. Assertion result might not hold. (<no position>)
  |
  = help: This could be caused by too small assertion timeout. Try increasing it by setting the configuration parameter ASSERT_TIMEOUT to a larger value.

error: [Prusti: verification error] precondition of pure function call might not hold.
  --> /home/data/eth/rust/prusti-dev/prusti-tests/tests/verify_overflow/pass/competitve_programming/dp/TheJackpot_10684_failing.rs:13:27
   |
13 | #[ensures(forall(|k: u32| check(k)))]
   |                           ^^^^^^^^
   |
note: the failing assertion is here
  --> /home/data/eth/rust/prusti-dev/prusti-tests/tests/verify_overflow/pass/competitve_programming/dp/TheJackpot_10684_failing.rs:10:12
   |
10 | #[requires(_x >= 0)]
   |            ^^^^^^^

Verification failed
error: aborting due to 2 previous errors; 3 warnings emitted
```

**Comment by fpoli** [comment]
We just renamed `CHECK_BINARY_OPERATIONS` to `CHECK_OVERFLOWS`. The internal error is anyway a bug and even with overflow checks enabled line 13 fails.

**Comment by fpoli** [comment]
The quantification fails because we don't (yet) encode value ranges for quantified variables.

**Comment by vakaras** [comment]
Fixed in #268
